optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsScoringBiConstraintStream.java

```diff
@@ -0,0 +1,161 @@
+        if (intMatchWeigher == null) {
+            throw new IllegalArgumentException("The matchWeigher (null) cannot be null.");
+        }
+    }
+
+    public DroolsScoringBiConstraintStream(DroolsConstraintFactory<Solution_> constraintFactory,
+            DroolsAbstractBiConstraintStream<Solution_, A, B> parent, DroolsConstraint<Solution_> constraint,
+            ToLongBiFunction<A, B> longMatchWeigher) {
+        this(constraintFactory, parent, constraint, false, null, longMatchWeigher, null);
+        if (longMatchWeigher == null) {
+            throw new IllegalArgumentException("The matchWeigher (null) cannot be null.");
+        }
+    }
+
+    public DroolsScoringBiConstraintStream(DroolsConstraintFactory<Solution_> constraintFactory,
+            DroolsAbstractBiConstraintStream<Solution_, A, B> parent, DroolsConstraint<Solution_> constraint,
+            BiFunction<A, B, BigDecimal> bigDecimalMatchWeigher) {
+        this(constraintFactory, parent, constraint, false, null, null, bigDecimalMatchWeigher);
+        if (bigDecimalMatchWeigher == null) {
+            throw new IllegalArgumentException("The matchWeigher (null) cannot be null.");
+        }
+    }
+
+    private DroolsScoringBiConstraintStream(DroolsConstraintFactory<Solution_> constraintFactory,
+            DroolsAbstractBiConstraintStream<Solution_, A, B> parent,
+            DroolsConstraint<Solution_> constraint, boolean noMatchWeigher, ToIntBiFunction<A, B> intMatchWeigher,
+            ToLongBiFunction<A, B> longMatchWeigher, BiFunction<A, B, BigDecimal> bigDecimalMatchWeigher) {
+        super(constraintFactory, parent);
+        this.constraint = constraint;
+        this.noMatchWeigher = noMatchWeigher;
+        this.intMatchWeigher = intMatchWeigher;
+        this.longMatchWeigher = longMatchWeigher;
+        this.bigDecimalMatchWeigher = bigDecimalMatchWeigher;
+    }
+
+    // ************************************************************************
+    // Pattern creation
+    // ************************************************************************
+
+    @Override
+    public void createRuleItemBuilders(List<RuleItemBuilder<?>> ruleItemBuilderList,
+            Global<? extends AbstractScoreHolder> scoreHolderGlobal) {
+        ruleItemBuilderList.add(getLeftPattern());
+        ruleItemBuilderList.add(getRightPattern());
+        Declaration<A> aVar = getLeftVariableDeclaration();
+        Declaration<B> bVar = getRightVariableDeclaration();
+        ConsequenceBuilder._3<? extends AbstractScoreHolder, A, B> consequence;
+        if (intMatchWeigher != null) {
+            consequence = PatternDSL.on(scoreHolderGlobal, aVar, bVar)
+                    .execute((drools, scoreHolder, a, b) -> {
+                        RuleContext kcontext = (RuleContext) drools;
+                        int weightMultiplier = intMatchWeigher.applyAsInt(a, b);
+                        scoreHolder.impactScore(kcontext, weightMultiplier);
```

Comment:
So in the end the decision was to use the `ScoreHolder` instead of `ScoreInliner`?

Reply:
Yep, Geoffrey couldn't make it work otherwise. We may revisit in the future.